### PR TITLE
Add unit tests for the PatrolTactic

### PR DESCRIPTION
**Pull request by amrm3lm**
### Description

- Added tests to patrolTactic, simulated the tactics' returned actions to completion in order to test the tactics transition logic from one patrol point to the next. 
- made small fix to patrolTactic as it was not returning new move actions after a robot reached a patrol point

### Resolved Issues
#1181 

### Review Checklist

**_It is the reviewers responsibility to also make sure every item here has been covered_**

- [x] **Function & Class comments**: All function definitions (usually in the `.h` file) should have a javadoc style comment at the start of them. For examples, see the functions defined in `thunderbots/software/geom`. Similarly, all classes should have an associated Javadoc comment explaining the purpose of the class.
- [x] **Remove all commented out code**
- [x] **Remove extra print statements**: for example, those just used for testing
- [x] **Resolve all TODO's**: All `TODO` (or similar) statements should either be completed or associated with a github issue
- [x] **Justify drops in code coverage**: If this PR results in a non-trivial drop in code coverage (a bot should post a coverage diagram as a comment), please justify why we can't test the code that's not covered.

### CODE REVIEW

**Review by garethellis0 — CHANGES_REQUESTED** [review]
Pretty clean, nice job! Just a few small comment

File: src/software/ai/hl/stp/tactic/patrol_tactic_test.cpp [hunk 1]
```diff
@@ -0,0 +1,289 @@
...
+    World world = ::Test::TestUtil::createBlankTestingWorld();
+
+    PatrolTactic tactic = PatrolTactic(patrolPoints, 1.0, Angle::zero(), 2.0);
+
+    tactic.updateRobot(assigned_robot);
+
+    double cost = tactic.calculateRobotCost(assigned_robot, world);
+
+    EXPECT_EQ(COST_OF_ASSIGNED_ROBOT, cost);
+}
```
Comment (kelvinkoon) [inline src/software/ai/hl/stp/tactic/patrol_tactic_test.cpp:306]: EOF extra line needed here.

File: src/software/ai/hl/stp/tactic/patrol_tactic_test.cpp [hunk 2]
```diff
@@ -0,0 +1,289 @@
...
+            auto expected_action = std::make_shared<MoveAction>(false);
+            expected_action->updateControlParams(
+                robot, patrolPoints[i], angle, speedAtPatrolPoints, DribblerEnable::OFF,
+                MoveType::NORMAL, AutokickType::NONE, BallCollisionType::AVOID);
+            expectedActions.push_back(expected_action);
+        }
+    }
+
+    /*
+     *  populates the robotStateToCompleteActions test variable in the order of patrol
```
Comment (kelvinkoon) [inline src/software/ai/hl/stp/tactic/patrol_tactic_test.cpp]: This comment style seems different to ones below such as "simulateActionToCompletion". It'd be best if they all adhered to a consistent style similar to simulateActionToCompletion where there is a section on what the function does and @params included.

File: src/software/ai/hl/stp/tactic/patrol_tactic_test.cpp (hunk 1 above)
Comment (kelvinkoon) [inline src/software/ai/hl/stp/tactic/patrol_tactic_test.cpp:306]: EOF extra line required here.

File: src/software/ai/hl/stp/tactic/patrol_tactic_test.cpp [hunk 3]
```diff
@@ -0,0 +1,289 @@
...
+        {
+            auto expected_action = std::make_shared<MoveAction>(false);
+            expected_action->updateControlParams(
+                robot, patrolPoints[i], angle, speedAtPatrolPoints, DribblerEnable::OFF,
+                MoveType::NORMAL, AutokickType::NONE, BallCollisionType::AVOID);
+            expectedActions.push_back(expected_action);
+        }
+    }
+
+    /*
```
Comment (kelvinkoon) [inline src/software/ai/hl/stp/tactic/patrol_tactic_test.cpp]: Function comment is inconsistent with the ones below. The 'simulateActionToCompletion' comment below including what the function does and @params included should be followed. Similar should be done for functions below (some functions have shorter comments or none at all).

**Review by kelvinkoon — CHANGES_REQUESTED** [review]
Overall, LGTM :100: 
Just a few minor comments below.

**Comment by garethellis0** [comment]
@amrm3lm - lemme know when you want another review on this!

**Comment by amrm3lm** [comment]
ready for re review @garethellis0 @kelvinkoon

File: src/software/ai/hl/stp/tactic/patrol_tactic_test.cpp [hunk 4]
```diff
@@ -0,0 +1,305 @@
...
+              AngularVelocity::zero(), Timestamp::fromSeconds(0));
+
+    std::vector<Point> patrol_points{Point(2, 2), Point(4, 3)};
+    World world = ::Test::TestUtil::createBlankTestingWorld();
+
+    PatrolTactic tactic = PatrolTactic(patrol_points, 1.0, Angle::zero(), 2.0);
+
+    double cost = tactic.calculateRobotCost(non_assigned_robot, world);
+
+    EXPECT_NEAR(0.15, cost, 0.05);
```
Comment (HaomiaoZ) [inline src/software/ai/hl/stp/tactic/patrol_tactic_test.cpp:282]: 0.15 here seems to be a bit magical to me. Just wondering why the cost is compared with 0.15 here?

**Review by HaomiaoZ — COMMENTED** [review]
Overall LGTM. I left a separate comment since I don't quite understand what is going on there.

**Review by garethellis0 — DISMISSED** [review]
LGTM (pending comments from others here) :+1: :tada: - Just as a note for the future, could you leave a comment reply to each of my comments? Even if it's as simple as "changed" or "not changed because...." it helps when I come back to re-review stuff.

File: src/software/ai/hl/stp/tactic/patrol_tactic_test.cpp (hunk 3 above)
Comment (amrm3lm) [reply src/software/ai/hl/stp/tactic/patrol_tactic_test.cpp]: added consistent comments to all functions

**Comment by amrm3lm** [comment]
> LGTM (pending comments from others here) +1 tada - Just as a note for the future, could you leave a comment reply to each of my comments? Even if it's as simple as "changed" or "not changed because...." it helps when I come back to re-review stuff.

yeah thats my bad, will reply to all comments

File: src/software/ai/hl/stp/tactic/patrol_tactic_test.cpp (hunk 1 above)
Comment (amrm3lm) [reply src/software/ai/hl/stp/tactic/patrol_tactic_test.cpp:306]: In my ide there is a line. Isnt it only needed if the small red symbol appears?

Comment (kelvinkoon) [reply src/software/ai/hl/stp/tactic/patrol_tactic_test.cpp:306]: Should be fine then.

**Review by kelvinkoon — DISMISSED** [review]
LGTM! :+1: Just seems like you need to fix formatting and should be good to go.

File: src/software/ai/hl/stp/tactic/patrol_tactic_test.cpp (hunk 4 above)
Comment (amrm3lm) [reply src/software/ai/hl/stp/tactic/patrol_tactic_test.cpp:282]: For a patrol tactic that is unassigned the cost of assigning a robot to the tactic is given by 

        `double cost = dist / world.field().totalXLength();`

at the bottom of PatrolTactic::calculateRobotCost().
 
`dist` is the distance between the robots current position (robotStartingPoint (1,1)) and the first patrol point (2,2), which is sqrt(2). 
`world.field().totalXLength();` is something set by the `test::TestUtil::createBlankTestingWorld();`  based on a real fields x length and some buffer. Its value is 9.6. 
Dividing the two values gives 0.1473 ~= 0.15

Hope thats a good enough answer :)

Comment (HaomiaoZ) [reply src/software/ai/hl/stp/tactic/patrol_tactic_test.cpp:282]: make sense, Thanks!

**Review by MathewMacDougall — APPROVED** [review]
LGTM :+1: :tada:

**Review by kelvinkoon — APPROVED** [review]

**Pull request merged**